workflows/background_check.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	checkID := workflow.GetInfo(w.ctx).WorkflowExecution.RunID
+
+	ctx := workflow.WithChildOptions(w.ctx, workflow.ChildWorkflowOptions{
+		WorkflowID: mappings.AcceptWorkflowID(checkID),
+	})
+	consentWF := workflow.ExecuteChildWorkflow(ctx, Accept, types.AcceptWorkflowInput{
+		Email:   email,
+		CheckID: checkID,
+	})
+	err = consentWF.Get(ctx, &r)
+
+	return r, err
+}
+
+func (w *backgroundCheckWorkflow) sendDeclineEmail(email string) error {
+	w.pushStatus(types.BackgroundCheckStatusDeclined)
+
+	ctx := workflow.WithActivityOptions(w.ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: time.Minute,
+	})
+
+	f := workflow.ExecuteActivity(ctx, a.SendDeclineEmail, types.SendDeclineEmailInput{Email: email})
+	return f.Get(w.ctx, nil)
+}
+
+func (w *backgroundCheckWorkflow) sendReportEmail(email string) error {
+	w.pushStatus(types.BackgroundCheckStatusCompleted)
+
+	ctx := workflow.WithActivityOptions(w.ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: time.Minute,
+	})
+
+	f := workflow.ExecuteActivity(ctx, a.SendReportEmail, types.SendReportEmailInput{Email: email, State: w.BackgroundCheckState})
+	return f.Get(ctx, nil)
+}
+
+func (w *backgroundCheckWorkflow) startCheck(name string, checkWorkflow interface{}, checkInputs ...interface{}) {
+	f := workflow.ExecuteChildWorkflow(
+		workflow.WithChildOptions(w.ctx, workflow.ChildWorkflowOptions{
+			WorkflowID: mappings.CheckWorkflowID(w.checkID, name),
+		}),
+		checkWorkflow,
+		checkInputs...,
+	)
+	w.checkFutures = append(w.checkFutures, f)
+	w.checkSelector.AddFuture(f, func(f workflow.Future) {
+		var result interface{}
+
+		err := f.Get(w.ctx, &result)
+		if err != nil {
+			w.logger.Error(fmt.Sprintf("%s search failed: %v", name, err))
+		}
+
+		w.Checks[name] = result
+	})
+}
+
+func (w *backgroundCheckWorkflow) waitForChecks() {
+	w.logger.Info(fmt.Sprintf("Waiting for %d checks", len(w.checkFutures)))
+
+	for i := 0; i < len(w.checkFutures); i++ {
+		w.checkSelector.Select(w.ctx)
+	}
+}
+
+func BackgroundCheck(ctx workflow.Context, input types.BackgroundCheckWorkflowInput) error {
+	w, err := newBackgroundCheckWorkflow(
+		ctx,
+		types.BackgroundCheckState{
+			Email: input.Email,
+			Tier:  input.Package,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	response, err := w.waitForAccept(w.Email)
+	if err != nil {
+		return err
+	}
+
+	w.Accepted = response.Accepted
+
+	if !w.Accepted {
+		return w.sendDeclineEmail(config.HiringManagerEmail)
+	}
+
+	w.CandidateDetails = response.CandidateDetails
+
+	err = w.pushStatus(types.BackgroundCheckStatusRunning)
+	if err != nil {
+		return err
+	}
+
+	ssnTrace := workflow.ExecuteChildWorkflow(
+		ctx,
+		SSNTrace,
+		types.SSNTraceWorkflowInput{FullName: w.CandidateDetails.FullName, SSN: w.CandidateDetails.SSN},
+	)
+
+	err = ssnTrace.Get(ctx, &w.SSNTrace)
+	if err != nil {
+		return err
+	}
+
+	if !w.SSNTrace.SSNIsValid || w.Tier != "full" {
+		return w.sendReportEmail(config.HiringManagerEmail)
+	}
+
+	w.startCheck(
```

Comment:
Why are these child workflows instead of simple activities? This seems overly complicated. Even ones like `StateCriminalSearch` can just make X number of HTTP calls in the activity.

Reply:
https://github.com/temporalio/background-checks/blob/8550c79073dfb6d03afdb1781cf8b588723e6c4d/docs/temporal-application-design-and-implementation.md#why-use-workflows-for-searches-instead-of-activities

Reply:
I am not seeing any of those benefits apply here.

A workflow that does nothing but execute an activity doesn't seem to have any of those benefits. I don't personally agree with it. You say division of responsibilities, but I say over-abstraction. Even if you may agree with the pattern in principle, you have added an intermediary to an HTTP call or two without any noticeable benefit.

I think we should work really hard on simplicity, code reduction, etc. A workflow that needs to execute a few activities, wait, etc doesn't need so many files, approaches, etc. Maybe the human input aspect is what's tripping me up and is not clear from the simple HTTP call.